### PR TITLE
feat: add span and transaction outcome

### PR DESCRIPTION
**Pull request by jahtalab**
Part of #1814
### Checklist

- [x] Implement code
- [x] Add tests
- [x] Update TypeScript typings
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/master/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -120,6 +120,8 @@ declare class Agent implements Taggable, StartSpanFn {
   registerMetric(name: string, labels: Labels, callback: Function): void;
```
Comment (astorm) [inline index.d.ts:120]: I see here we've created a new type, `Outcome`, that can be one of three string values.  We've also used the type to add a new `outcome` property to the `GenericSpan` class.

File: lib/instrumentation/http-shared.js [hunk 2]
```diff
@@ -206,6 +206,15 @@ exports.traceOutgoingRequest = function (agent, moduleName, method) {
...
         })
+
+        if (typeof statusCode !== 'undefined') {
+          if (statusCode >= 400) {
+            span.outcome = 'failure'
+          } else {
+            span.outcome = 'success'
+          }
+        }
+
```
Comment (astorm) [inline lib/instrumentation/http-shared.js:217]: Here I see we're setting the `outcome` property of a span based on the the HTTP status code -- anything greater than 400 is a `failure`. This matches the behavior described in the spec.  The inline type checking seems prudent given how the lifecycle of the `statusCode` variable is a little weird in this module.

File: lib/instrumentation/modules/http.js [hunk 3]
```diff
@@ -41,6 +41,11 @@ module.exports = function (http, agent, { enabled }) {
         // It shouldn't be possible for the statusCode to be falsy, but just in
         // case we're in a bad state we should avoid throwing
         trans.result = 'HTTP ' + (this.statusCode || '').toString()[0] + 'xx'
+        if (this.statusCode >= 500) {
+          trans.outcome = 'failure'
+        } else {
+          trans.outcome = 'success'
+        }
```
Comment (astorm) [inline lib/instrumentation/modules/http.js:49]: Here I see that during our handling/wrapping of the `http.ServerResponse` method we're examining the status code used in the response, and if it's greater than or equal to 500 we set the outcome as a failure -- otherwise we set the outcome to success.  This matches the behavior described in the spec.

File: lib/instrumentation/modules/http2.js [hunk 4]
```diff
@@ -80,6 +80,11 @@ module.exports = function (http2, agent, { enabled }) {
       var status = headers[':status'] || 200
       trans.result = 'HTTP ' + status.toString()[0] + 'xx'
       trans.res.statusCode = status
+      if (status >= 500) {
+        trans.outcome = 'failure'
+      } else {
+        trans.outcome = 'success'
+      }
       trans.res.headers = mergeHeaders(trans.res.headers, headers)
```
Comment (astorm) [inline lib/instrumentation/modules/http2.js:88]: Here I see that during our handling of an http2 response we're setting the transaction's outcome to `failure` if the http status code is greater than or equal to 500.  This behavior matches the spec.

File: lib/instrumentation/modules/http2.js [hunk 5]
```diff
@@ -179,6 +184,14 @@ module.exports = function (http2, agent, { enabled }) {
...
         })
 
+        if (typeof statusCode !== 'undefined') {
+          if (statusCode >= 400) {
+            span.outcome = 'failure'
+          } else {
+            span.outcome = 'success'
+          }
+        }
+
```
Comment (astorm) [inline lib/instrumentation/modules/http2.js:194]: Here I see that during our handling of calls the the `request` method of of objects returned by http.connect we're examining the HTTP2 response code and setting the outcome property to `failure` if the status code is greater than or equal to 400.  This behavior matches the spec.

File: lib/instrumentation/span.js [hunk 6]
```diff
@@ -153,7 +154,8 @@ Span.prototype._encode = function (cb) {
       duration: self.duration(),
       context: undefined,
       stacktrace: frames,
-      sync: self.sync
+      sync: self.sync,
+      outcome: self.outcome
     }
```
Comment (astorm) [inline lib/instrumentation/span.js:159]: Here I see we're adding `outcome` to the serialized span payload.

File: lib/instrumentation/span.js [hunk 7]
```diff
@@ -36,6 +36,7 @@ function Span (transaction, name, ...args) {
 
   this.transaction = transaction
   this.name = name || 'unnamed'
+  this.outcome = 'unknown'
```
Comment (astorm) [inline lib/instrumentation/span.js:40]: 👍 for a default outcome of unknown.

File: lib/instrumentation/span.js [hunk 8]
```diff
@@ -153,7 +154,8 @@ Span.prototype._encode = function (cb) {
       duration: self.duration(),
       context: undefined,
       stacktrace: frames,
-      sync: self.sync
+      sync: self.sync,
+      outcome: self.outcome
```
Comment (astorm) [inline lib/instrumentation/span.js:158]: @jahtalab What is the `sync` property here for?  Is this related to the `outcome` functionality?

File: lib/instrumentation/transaction.js [hunk 9]
```diff
@@ -129,7 +130,8 @@ Transaction.prototype.toJSON = function () {
     sync: this.sync,
     span_count: {
       started: this._builtSpans
-    }
+    },
+    outcome: this.outcome
   }
```
Comment (astorm) [inline lib/instrumentation/transaction.js:135]: I see here we're adding the outcome to the serialized transaction.

**Review by astorm — COMMENTED** [review]
@jahtalab this looks like a great and also showed me where a few of the important agent hooking/listening points were with regards to incoming and outgoing HTTP transactions.  TIL :) 

I had one question about the `sync` property below -- the rest of my review is just stating out loud what I see in order to make sure we're on the same page.  

In addition to that sync question -- how would you feel about creating a `./lib/errors` (or similar) module so logic for handling status code is in one central place?  We don't need to do the full API portions of the spec right now to land this, but unless there's a compelling reason not to I'd rather have the logic for identifying what status codes relate to which outcomes in one place.  It'll save us a bunch of time/headache when we get around to implementing an API and configuration based approach for setting this value.

File: lib/instrumentation/span.js [hunk 10]
```diff
@@ -36,6 +36,7 @@ function Span (transaction, name, ...args) {
 
   this.transaction = transaction
   this.name = name || 'unnamed'
+  this.outcome = 'unknown'
```
Comment (vigneshshanmugam) [inline lib/instrumentation/span.js:39]: can you move this to `GenericsSpan` class as `outcome` belongs to both Transaction and Span.

File: test/instrumentation/modules/http/outgoing.js [hunk 11]
```diff
@@ -230,6 +231,10 @@ function abortTest (type, handler) {
         t.equal(data.transactions.length, 1, 'has one transaction')
         t.equal(data.spans.length, 1, 'has one span')
         t.equal(data.spans[0].name, `${req.method} localhost:${port}/`, 'has expected span name')
+        if (typeof httpContext.status_code === 'undefined') {
```
Comment (vigneshshanmugam) [inline test/instrumentation/modules/http/outgoing.js:234]: can you add else case which also checks for partial response? 

https://github.com/elastic/apm-agent-nodejs/blob/5bd3eb79fd806befb52c87bed1e1883229156918/test/instrumentation/modules/http/outgoing.js#L59-L71

**Review by vigneshshanmugam — APPROVED** [review]
LGTM 👍  with moving the `outcome` to GenericSpan.

**Comment by jahtalab** [comment]
@astorm , Thanks for the review, I agree with refactoring the logic related to the outcome into a central place 👍 . I've now added `setOutcome` function on Span/Transaction that contains this logic. I think since the `outcome` logic depends on event type (span/transaction), and also it changes its internal state, it would be better to keep the logic in their respective events. Let me know WDYT?

In regards to the other comments, your observations are aligned with what I intended 👍, (I'll mark them as resolved)

File: lib/instrumentation/http-shared.js (hunk 2 above)
Comment (jahtalab) [reply lib/instrumentation/http-shared.js:210]: I've add a comment in `span.setOutcome` with regards to the type checking to make it clear

File: lib/instrumentation/span.js (hunk 8 above)
Comment (jahtalab) [reply lib/instrumentation/span.js:172]: The `sync` property is not related to outcome in any way. The change here only adds a comma `,`, is that what you meant?

Comment (astorm) [reply lib/instrumentation/span.js:172]: Ah -- I had misread the diff.  We're all good here.

**Review by astorm — APPROVED** [review]
💥 Looks good, thanks for making those changes.  Approving and merging.
